swiftstorageprovider/src/main/java/org/duracloud/swiftstorage/SwiftStorageProvider.java

```diff
@@ -0,0 +1,119 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides content storage backed by OpenStack Swift with S3 API middleware.
+ *
+ * @author Andy Foster
+ * Date: Feb 25, 2019
+ */
+public class SwiftStorageProvider extends S3StorageProvider {
+
+    private final Logger log = LoggerFactory.getLogger(SwiftStorageProvider.class);
+
+    public SwiftStorageProvider(String accessKey, String secretKey, Map<String, String> options) {
+        super(accessKey, secretKey, options);
+    }
+
+    public SwiftStorageProvider(AmazonS3 s3Client, String accessKey) {
+        super(s3Client, accessKey, null);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public StorageProviderType getStorageProviderType() {
+        return StorageProviderType.SWIFT_S3;
+    }
+
+    @Override
+    protected Bucket createBucket(String spaceId) {
+        String bucketName = getNewBucketName(spaceId);
+        try {
+            Bucket bucket = s3Client.createBucket(bucketName);
+
+            // Swift has no concept of bucket lifecycle
+
+            return bucket;
+        } catch (AmazonClientException e) {
+            String err = "Could not create S3 bucket with name " + bucketName
+                         + " due to error: " + e.getMessage();
+            throw new StorageException(err, e, RETRY);
+        }
+    }
+
+    // Swift access keys are longer than 20 characters, and creating
+    // a bucket starting with your access key causes problems.
+    @Override
+    protected String getNewBucketName(String spaceId) {
+        String truncatedKey = truncateKey(accessKeyId);
+        return S3ProviderUtil.createNewBucketName(truncatedKey, spaceId);
+    }
+
+    @Override
+    protected String getSpaceId(String bucketName) {
+        String spaceId = bucketName;
+        String truncatedKey = truncateKey(accessKeyId);
+        if (isSpace(bucketName)) {
+            spaceId = spaceId.substring(truncatedKey.length() + 1);
+        }
+        return spaceId;
+    }
+
+    @Override
+    protected Map<String, String> getAllSpaceProperties(String spaceId) {
+        log.debug("getAllSpaceProperties(" + spaceId + ")");
+
+        // Will throw if bucket does not exist
+        String bucketName = getBucketName(spaceId);
+
+        // Retrieve space properties from bucket tags
+        // But Swift has no concept of tags, so just return a map with the
+        // space count.
+        Map<String, String> spaceProperties = new HashMap<>();
+
+        // Add space count
+        spaceProperties.put(PROPERTIES_SPACE_COUNT,
+                            getSpaceCount(spaceId, MAX_ITEM_COUNT));
+
+        return spaceProperties;
+    }
+
+    @Override
+    protected void doSetSpaceProperties(String spaceId,
+                                        Map<String, String> spaceProperties) {
+        log.debug("setSpaceProperties(" + spaceId + ")");
+
+        // Swift has no concept of tags, so do nothing.
```

Comment:
We'll need to consider how to handle this, as space-level access control settings are saved as space properties. Without some way to capture that information only administrative users will be able to access spaces.